x/incentive/keeper/rewards.go

```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
+		}
+
+		// Delegators don't accumulate rewards if their validator is unbonded/slashed
+		if validator.GetStatus() != sdk.Bonded {
+			continue
+		}
+
+		if validator.GetTokens().IsZero() {
+			continue
+		}
+
+		sharesToTokens := sdk.NewDecFromInt(validator.GetTokens()).Quo(validator.GetDelegatorShares())
+		delegatedTokens := sharesToTokens.Mul(delegation.GetShares())
+		if delegatedTokens.IsZero() || delegatedTokens.IsNegative() {
+			continue
+		}
+
+		totalDelegated = totalDelegated.Add(delegatedTokens)
+		// TODO: do we need to consider respective share of validator tokens?
+		// delegationShare := delegationTokens.Quo(sdk.NewDecFromInt(totalBonded))
+		// rewardsEarned := delegationShare.Mul(sdk.NewDecFromInt(rewardsToDistribute)).RoundInt()
+	}
+
+	rewardsEarned := rewardsAccumulatedFactor.Mul(totalDelegated).RoundInt()
+	if rewardsEarned.IsZero() || rewardsEarned.IsNegative() {
+		return
+	}
+
+	// Add rewards to delegator's hard claim
+	newRewardsCoin := sdk.NewCoin(types.HardLiquidityRewardDenom, rewardsEarned)
+	claim.Reward = claim.Reward.Add(newRewardsCoin)
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// AccumulateHardDelegatorRewards updates the rewards accumulated for the input reward period
+func (k Keeper) AccumulateHardDelegatorRewards(ctx sdk.Context, delAddr sdk.AccAddress, rewardPeriod types.RewardPeriod) error {
+	previousAccrualTime, found := k.GetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType)
+	if !found {
+		k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+	timeElapsed := CalculateTimeElapsed(rewardPeriod, ctx.BlockTime(), previousAccrualTime)
+	if timeElapsed.IsZero() {
+		return nil
+	}
+	if rewardPeriod.RewardsPerSecond.Amount.IsZero() {
+		k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+
+	totalBonded := k.stakingKeeper.TotalBondedTokens(ctx).ToDec()
+	if totalBonded.IsZero() {
+		k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+
+	newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
+	delegatorFactor, found := k.hardKeeper.GetDelegatorInterestFactor(ctx, rewardPeriod.CollateralType)
+	if !found {
+		k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+	rewardFactor := newRewards.ToDec().Mul(delegatorFactor).Quo(totalBonded)
+
+	previousRewardFactor, found := k.GetHardDelegatorRewardFactor(ctx, rewardPeriod.CollateralType)
+	if !found {
+		previousRewardFactor = sdk.ZeroDec()
+	}
+	newRewardFactor := previousRewardFactor.Add(rewardFactor)
+	k.SetHardDelegatorRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
+	k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+	return nil
+}
+
+// TODO: could a single delegation update the delegator index, effectively updating the value before
```

Comment:
Not sure what scenario this is referring to. Can you specify it a bit more?

Reply:
Ah this TODO is outdated, removed in https://github.com/Kava-Labs/kava/pull/776/commits/3e33ae0354f7e0e1f658db8c9fa6e38bf5e1dc91.